File: 0002-add-two-numbers/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* ins(int &rem, ListNode* sum){
        ListNode* curr=new ListNode();
        sum->next=curr;
        if(rem<10){
            curr->val=rem;
            rem=0;
        }
        else{
            curr->val=(rem%10);
            rem/=10;
        }
        return curr;
    }

    void result(int rem, ListNode* l1, ListNode* l2, ListNode* sum){
        if(l1==nullptr&&l2==nullptr){
            if(rem){
                ListNode* curr=ins(rem,sum);
                result(rem,l1,l2,curr);
            }
        }else if(l1==nullptr){
            rem=rem+l2->val;
            ListNode* curr=ins(rem,sum);
            result(rem,l1,l2->next,curr);
        }else if(l2==nullptr){
            rem=rem+l1->val;
            ListNode* curr=ins(rem,sum);
            result(rem,l1->next,l2,curr);
        }else{
            rem=rem+l1->val+l2->val;
            ListNode* curr=ins(rem,sum);
            result(rem,l1->next,l2->next,curr);
        }
    }
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* sum=new ListNode();
        result(0,l1,l2,sum);
        return sum->next;
    }
};

File: 0002-add-two-numbers/solution.go
func ins(rem int, sum *ListNode) (int, *ListNode) {
    curr := &ListNode{}
    sum.Next = curr
    if rem < 10 {
        curr.Val = rem
        rem = 0
    } else {
        curr.Val = (rem % 10)
        rem /= 10
    }
    return rem, curr
}

func result(rem int, l1 *ListNode, l2 *ListNode, sum *ListNode) {
    if l1 == nil && l2 == nil {
        if rem != 0 {
            rem, curr := ins(rem, sum)
            result(rem, l1, l2, curr)
        }
    } else if l1 == nil {
        rem = rem + l2.Val
        rem, curr := ins(rem, sum)
        result(rem, l1, l2.Next, curr)
    } else if l2 == nil {
        rem = rem + l1.Val
        rem, curr := ins(rem, sum)
        result(rem, l1.Next, l2, curr)
    } else {
        rem = rem + l1.Val + l2.Val
        rem, curr := ins(rem, sum)
        result(rem, l1.Next, l2.Next, curr)
    }
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    sum := &ListNode{}
    result(0, l1,l2, sum)
    return sum.Next
}
